src/components/sticker.rs

```diff
@@ -0,0 +1,191 @@
+
+mod imp {
+    use super::*;
+    use std::cell::{Cell, RefCell};
+
+    #[derive(Default)]
+    pub struct Sticker {
+        pub(super) message_id: Cell<i64>,
+        pub(super) aspect_ratio: Cell<f64>,
+        pub(super) max_size: Cell<i32>,
+
+        pub(super) child: RefCell<Option<gtk::Widget>>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Sticker {
+        const NAME: &'static str = "ComponentsSticker";
+        type Type = super::Sticker;
+        type ParentType = gtk::Widget;
+    }
+
+    impl ObjectImpl for Sticker {}
+
+    impl WidgetImpl for Sticker {
+        fn measure(&self, orientation: gtk::Orientation, _for_size: i32) -> (i32, i32, i32, i32) {
+            let size = self.max_size.get();
+            let aspect_ratio = self.aspect_ratio.get();
+
+            let min_size = 1;
+
+            let size = if let gtk::Orientation::Horizontal = orientation {
+                if aspect_ratio >= 1.0 {
+                    size
+                } else {
+                    (size as f64 * aspect_ratio) as i32
+                }
+            } else if aspect_ratio >= 1.0 {
+                (size as f64 / aspect_ratio) as i32
+            } else {
+                size
+            }
+            .max(min_size);
+
+            (size, size, -1, -1)
+        }
+
+        fn size_allocate(&self, width: i32, height: i32, baseline: i32) {
+            if let Some(child) = &*self.child.borrow() {
+                child.allocate(width, height, baseline, None);
+            }
+        }
+
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            if let Some(child) = &*self.child.borrow() {
+                self.obj().snapshot_child(child, snapshot);
+            }
+        }
+
+        fn request_mode(&self) -> gtk::SizeRequestMode {
+            gtk::SizeRequestMode::HeightForWidth
+        }
+    }
+}
+
+glib::wrapper! {
+    pub struct Sticker(ObjectSubclass<imp::Sticker>)
+        @extends gtk::Widget;
+}
+
+impl Sticker {
+    pub(crate) fn set_sticker(
+        &self,
+        sticker: TdSticker,
+        looped: bool,
+        session: Session,
+        message_id: i64,
+    ) {
+        // TODO: draw sticker outline with cairo
+        self.set_child(None);
+
+        let imp = self.imp();
+
+        let aspect_ratio = sticker.width as f64 / sticker.height as f64;
+        imp.aspect_ratio.set(aspect_ratio);
+
+        imp.message_id.set(message_id);
+
+        let format = sticker.format;
+
+        spawn(clone!(@weak self as obj, @weak session => async move {
+            if sticker.sticker.local.is_downloading_completed {
+                obj.load_sticker(&sticker.sticker.local.path, looped, format).await;
+            } else {
+            let file_id = sticker.sticker.id;
+                obj.download_sticker(file_id, &session, looped, format).await
+            }
+        }));
+    }
+
+    pub(crate) fn set_max_size(&self, size: i32) {
+        self.imp().max_size.set(size);
+    }
```

Comment:
It would be good to have this as a property. Also, you probably want to [queue a resize](https://docs.gtk.org/gtk4/method.Widget.queue_resize.html) when this value is changed.

Also, I know what this property is for, but the name is a bit unclear. It may seem that this widget can resize and this sets the max size of the widget, but this widget is meant to have a constant size. I think a more clear name could be `longer-side-size`.